test/py_tb/mock_host.py

```diff
@@ -448,88 +495,201 @@ def get_net_ifc_tx_data_from_nic_blocking(self):
+            s.setsockopt(socket.IPPROTO_IP, socket.IP_HDRINCL, 1)
+        except socket.error as msg:
             import sys
-            sys.stdout.flush()
-            while True:
-                packet, (_ip, _port) = s.recvfrom(5120)
-                print("received packet:{}", packet)
-                import sys
-                sys.stdout.flush()
-                eth_frame = Ether(src=other_mac, dst=nic_mac, type=0x0800)
-                packet = bytes(eth_frame) + packet
-                for frag in split_packet_into_fragments(packet):
-                    nic.put_net_ifc_rx_data_to_nic(frag)
-
-        raw_socket_thread = threading.Thread(target=_send_proxy)
-        raw_socket_thread.start()
-        raw_socket_thread = threading.Thread(target=_receive_proxy)
-        raw_socket_thread.start()
-
-    @staticmethod
-    def connect_two_card(nic_a, nic_b):
-        def _forward_a():
-            # client_to_buf = dict()
-            while True:
-                data = nic_a.get_net_ifc_tx_data_from_nic_blocking()
-                nic_b.put_net_ifc_rx_data_to_nic(data)
+            sys.stderr.write('Socket could not be created. Error Code : ' +
+                             str(msg) + ' Message ')
+        s.bind((addr, port))
+        while not self.stop_flag:
+            packet, (_ip, _port) = s.recvfrom(5120)
+            eth_frame = Ether(src=remote_mac, dst=local_mac, type=0x0800)
+            packet = bytes(eth_frame) + packet
+            for frag in self.split_packet_into_fragments(packet):
+                self.pending_network_packet_rx.append(frag)
+                self.pending_network_packet_rx_sema.release()
+        s.close()
+
+    def _send_proxy(self):
+        try:
+            s = socket.socket(
+                socket.AF_INET, socket.SOCK_RAW, socket.IPPROTO_UDP)
+            s.setsockopt(socket.IPPROTO_IP, socket.IP_HDRINCL, 1)
+        except socket.error as msg:
+            import sys
+            sys.stderr.write('Socket could not be created. Error Code : ' +
+                             str(msg) + ' Message ')
+        data_buf = dict()
+        while not self.stop_flag:
+            self.pending_network_packet_tx_sema.acquire()
+            if self.stop_flag:
+                break
+            data = self.pending_network_packet_tx.popleft()
+            bitmask = struct.unpack('<Q', bytes(data.payload.byte_en))[0]
+            payload = self.get_data_with_bitmap(
+                bytes(data.payload.data), bitmask)
+            if data.header.tag not in data_buf:
+                data_buf[data.header.tag] = payload
+            else:
+                data_buf[data.header.tag] += payload
+            if data.payload.is_last:
+                packet = data_buf[data.header.tag]
+                dst_ip = socket.inet_ntoa(packet[30:34])
+                s.sendto(data_buf[data.header.tag][14:], (dst_ip, 0))
+                del data_buf[data.header.tag]
+        s.close()
+
+    # Generated by python
```

Comment:
see here  https://stackoverflow.com/questions/8871204/count-number-of-1s-in-binary-representation

Reply:
Yes, I have looked at it. There are mainly 4 methods here:
* using a lookup table (the method I currently using:https://stackoverflow.com/a/8871279)
* using i & -i trick
* using some mask tricks(like https://stackoverflow.com/a/8871435). But it seems they are most for int 
* using Python builtin library(`bin(num).count('1')`)

Actually, I used to use the second way in the last version( i & -i). Should I try the third way?

Reply:
The loopup table way is too verbose. I think the other 3 methods are all fine. You said that mask tricks is for int, but byteEn must be a int, isn't it? so what's your concern about? the builtin libary is the simplist way I think.

Reply:
I mean the 32-bit `int` in C/C++, not the type `int` in `Python`. I tried other tricks for 64-bits in C,  but they didn't work well. I'd like to use the built-in library method instead.